src/library/scala/concurrent/impl/Promise.scala

```diff
@@ -7,400 +7,441 @@
-        case r: Try[_]          => runnable.executeWithValue(r.asInstanceOf[Try[T]])
-        case dp: DefaultPromise[_] => compressedRoot(dp).dispatchOrAddCallback(runnable)
-        case listeners: List[_] => if (compareAndSet(listeners, runnable :: listeners)) ()
-                                   else dispatchOrAddCallback(runnable)
+    @tailrec private final def dispatchOrAddCallbacks[C <: Callbacks[T]](state: AnyRef, callbacks: C): C =
+      if (state.isInstanceOf[Try[T]]) {
+        callbacks.submitWithValue(state.asInstanceOf[Try[T]])
+        callbacks
+      } else if (state.isInstanceOf[Callbacks[T]]) {
+        val newCallbacks = if (state eq Noop) callbacks else new ManyCallbacks[T](callbacks, state.asInstanceOf[Callbacks[T]])
+        if(compareAndSet(state, newCallbacks)) callbacks
+        else dispatchOrAddCallbacks(get(), callbacks)
+      } else /*if (state.isInstanceOf[Link[T]])*/ {
+        val p = state.asInstanceOf[Link[T]].promise(this)
+        p.dispatchOrAddCallbacks(p.get(), callbacks)
       }
-    }
 
-    /** Link this promise to the root of another promise using `link()`. Should only be
-     *  be called by transformWith.
-     */
-    protected[concurrent] final def linkRootOf(target: DefaultPromise[T]): Unit = link(target.compressedRoot())
-
-    /** Link this promise to another promise so that both promises share the same
-     *  externally-visible state. Depending on the current state of this promise, this
-     *  may involve different things. For example, any onComplete listeners will need
-     *  to be transferred.
-     *
-     *  If this promise is already completed, then the same effect as linking -
-     *  sharing the same completed value - is achieved by simply sending this
-     *  promise's result to the target promise.
+    /** Link this promise to the root of another promise.
      */
-    @tailrec
-    private def link(target: DefaultPromise[T]): Unit = if (this ne target) {
-      get() match {
-        case r: Try[_] =>
-          if (!target.tryComplete(r.asInstanceOf[Try[T]]))
+    @tailrec protected[concurrent] final def linkRootOf(target: DefaultPromise[T], link: Link[T]): Unit =
+      if (this ne target) {
+        val state = get()
+        if (state.isInstanceOf[Try[T]]) {
+          if(!target.tryComplete(state.asInstanceOf[Try[T]]))
             throw new IllegalStateException("Cannot link completed promises together")
-        case dp: DefaultPromise[_] =>
-          compressedRoot(dp).link(target)
-        case listeners: List[_] if compareAndSet(listeners, target) =>
-          if (listeners.nonEmpty)
-            listeners.asInstanceOf[List[CallbackRunnable[T]]].foreach(target.dispatchOrAddCallback(_))
-        case _ =>
-          link(target)
+        } else if (state.isInstanceOf[Link[T]]) {
+          // TODO: does relinking ever make sense?
+          state.asInstanceOf[Link[T]].relink(link = state.asInstanceOf[Link[T]], target = target, owner = this)
```

Comment:
@richdougherty Rich, I'd love to get your input on this. Does re-linking ever make sense really? Callbacks have already been migrated, so re-pointing it to somewhere else seems like a problem?

Reply:
Ping @richdougherty

Reply:
I've pasted the old comments below. Is this new relinking code the same as the old code which breaks/compresses long chains of promises and gets them all to point at a root promise?

```
-   * The problem of leaks is solved by automatically breaking these chains of
-   * promises, so that promises don't refer to each other in a long chain. This
-   * allows each promise to be individually collected. The idea is to "flatten"
-   * the chain of promises, so that instead of each promise pointing to its
-   * neighbour, they instead point directly the promise at the root of the
-   * chain. This means that only the root promise is referenced, and all the
-   * other promises are available for garbage collection as soon as they're no
-   * longer referenced by user code. 
...
-   * To mitigate the problem of the root promise changing, whenever a promise's
-   * methods are called, and it needs a reference to its root promise it calls
-   * the `compressedRoot()` method. This method re-scans the promise chain to
-   * get the root promise, and also compresses its links so that it links
-   * directly to whatever the current root promise is. This ensures that the
-   * chain is flattened whenever `compressedRoot()` is called. And since
-   * `compressedRoot()` is called at every possible opportunity (when getting a
-   * promise's value, when adding an onComplete handler, etc), this will happen
-   * frequently. Unfortunately, even this eager relinking doesn't absolutely
-   * guarantee that the chain will be flattened and that leaks cannot occur.
-   * However eager relinking does greatly reduce the chance that leaks will
-   * occur.
```

Is the relinking here responsible for "compressing" the links

Reply:
@richdougherty It's not the same, but it is based on that code. The new implementation correctly unlinks in case of completion, to not leave links post completion.

Reply:
@richdougherty But still: Does switching one link to another really make sense? Callbacks have already been migrated to another promise.

Reply:
In which cases would links remain chained together?

If there's a chance of building a long (possibly infinite) chain of links then that might be worth addressing.

Does the length of the chain affect the performance of adding a callback to the end of the chain? I.e. does the callback logic ever have to traverse a long chain? If so, it would be quicker to keep the chain short.

Reply:
@richdougherty Perhaps we're talking past each other. I'm talking about the case where a Future is linked to one logical chain and then gets linked to *another* logical chain. I.e. linkRootOf being invoked twice on the same DefaultPromise, but for different logical chains.

My question is: should it only be possible to call linkRootOf *once* for a DefaultPromise?

Reply:
In the original code `linkRootOf` could be invoked twice.

```scala
// from old code
def transformWith[S](f: Try[T] => Future[S])(implicit executor: ExecutionContext): Future[S] = {
  val p = new DefaultPromise[S]()
  onComplete { 
    ...
    case dp: DefaultPromise[_] => dp.asInstanceOf[DefaultPromise[S]].linkRootOf(p)
    ...
  }
  p.future
}

// example of linking twice
val rootPromise = Promise()
val root = rootPromise.future // Actually a DefaultPromise

fut1.transformWith(_ => root) // root.linkRootOf(...) call #1
fut2.transformWith(_ => root) // root.linkRootOf(...) call #2
```